From 752479262b3155653f1fe455452a7a0992f1adc3 Mon Sep 17 00:00:00 2001
From: Maladau Mou <maladau.mou@intel.com>
Date: Fri, 20 Mar 2015 20:27:37 +0000
Subject: [PATCH] spi/pxa2xx: add support for Cougar Mountain SPI controllers

Note: We re-use the DMA channel assignments from BSW,
but we have one less port than BSW does, so assume re-use of PORT_BSW0/BSW1.

Signed-off-by: Maladau Mou <maladau.mou@intel.com>
---
 drivers/spi/spi-pxa2xx-pci.c |   25 +++++++++++++++++++++++++
 drivers/spi/spi-pxa2xx.c     |    5 +++++
 2 files changed, 30 insertions(+)

diff --git a/drivers/spi/spi-pxa2xx-pci.c b/drivers/spi/spi-pxa2xx-pci.c
index b285294..25e9660 100644
--- a/drivers/spi/spi-pxa2xx-pci.c
+++ b/drivers/spi/spi-pxa2xx-pci.c
@@ -19,6 +19,8 @@ enum {
 	PORT_BSW0,
 	PORT_BSW1,
 	PORT_BSW2,
+	PORT_CGM0,
+	PORT_CGM1,
 };
 
 struct pxa_spi_info {
@@ -42,6 +44,11 @@ static struct dw_dma_slave bsw1_rx_param = { .src_id = 7 };
 static struct dw_dma_slave bsw2_tx_param = { .dst_id = 8 };
 static struct dw_dma_slave bsw2_rx_param = { .src_id = 9 };
 
+static struct dw_dma_slave cgm0_tx_param = { .dst_id = 0 };
+static struct dw_dma_slave cgm0_rx_param = { .src_id = 1 };
+static struct dw_dma_slave cgm1_tx_param = { .dst_id = 6 };
+static struct dw_dma_slave cgm1_rx_param = { .src_id = 7 };
+
 static bool lpss_dma_filter(struct dma_chan *chan, void *param)
 {
 	struct dw_dma_slave *dws = param;
@@ -92,6 +99,22 @@ static struct pxa_spi_info spi_info_configs[] = {
 		.tx_param = &bsw2_tx_param,
 		.rx_param = &bsw2_rx_param,
 	},
+	[PORT_CGM0] = {
+		.type = LPSS_SSP,
+		.port_id = 0,
+		.num_chipselect = 1,
+		.max_clk_rate = 50000000,
+		.tx_param = &cgm0_tx_param,
+		.rx_param = &cgm0_rx_param,
+	},
+	[PORT_CGM1] = {
+		.type = LPSS_SSP,
+		.port_id = 1,
+		.num_chipselect = 1,
+		.max_clk_rate = 50000000,
+		.tx_param = &cgm1_tx_param,
+		.rx_param = &cgm1_rx_param,
+	},
 };
 
 static int pxa2xx_spi_pci_probe(struct pci_dev *dev,
@@ -196,6 +219,8 @@ static const struct pci_device_id pxa2xx_spi_pci_devices[] = {
 	{ PCI_VDEVICE(INTEL, 0x228e), PORT_BSW0 },
 	{ PCI_VDEVICE(INTEL, 0x2290), PORT_BSW1 },
 	{ PCI_VDEVICE(INTEL, 0x22ac), PORT_BSW2 },
+	{ PCI_VDEVICE(INTEL, 0x2b8e), PORT_CGM0 },
+	{ PCI_VDEVICE(INTEL, 0x2b90), PORT_CGM1 },
 	{ },
 };
 MODULE_DEVICE_TABLE(pci, pxa2xx_spi_pci_devices);
diff --git a/drivers/spi/spi-pxa2xx.c b/drivers/spi/spi-pxa2xx.c
index 80f5d81..c2e2922 100644
--- a/drivers/spi/spi-pxa2xx.c
+++ b/drivers/spi/spi-pxa2xx.c
@@ -1152,6 +1152,10 @@ static struct pxa2xx_spi_acpi_desc cht_spi_desc = {
 	.cs_num = 2,
 };
 
+static struct pxa2xx_spi_acpi_desc cgm_spi_desc = {
+	.cs_num = 2,
+};
+
 static struct acpi_device_id pxa2xx_spi_acpi_match[] = {
 	{ "INT33C0", 0 },
 	{ "INT33C1", 0 },
@@ -1159,6 +1163,7 @@ static struct acpi_device_id pxa2xx_spi_acpi_match[] = {
 	{ "INT3431", 0 },
 	{ "80860F0E", (kernel_ulong_t)&byt_spi_desc },
 	{ "8086228E", (kernel_ulong_t)&cht_spi_desc },
+	{ "80862B8E", (kernel_ulong_t)&cgm_spi_desc },
 	{ },
 };
 MODULE_DEVICE_TABLE(acpi, pxa2xx_spi_acpi_match);
-- 
1.7.9.5

